File: frontend/package/src/Score/Pitch/Accidental.ts
enum Accidental {
  NAT = 0,
  SHARP = 1,
  DBLSHARP = 2,
  FLAT = -1,
  DBLFLAT = -2,
}

namespace Accidental {
  export function raise(a: Accidental): Accidental {
    switch (a) {
      case Accidental.DBLFLAT:
        return Accidental.FLAT;
      case Accidental.FLAT:
        return Accidental.NAT;
      case Accidental.NAT:
        return Accidental.SHARP;
      case Accidental.SHARP:
        return Accidental.DBLSHARP;
      default:
        return a;
    }
  }

  export function lower(a: Accidental): Accidental {
    switch (a) {
      case Accidental.FLAT:
        return Accidental.DBLFLAT;
      case Accidental.NAT:
        return Accidental.FLAT;
      case Accidental.SHARP:
        return Accidental.NAT;
      case Accidental.DBLSHARP:
        return Accidental.SHARP;
      default:
        return a;
    }
  }

  export function toString(a: Accidental): string {
    switch (a) {
      case Accidental.NAT:
        return '';
      case Accidental.SHARP:
        return '#';
      case Accidental.DBLSHARP:
        return '##';
      case Accidental.FLAT:
        return 'b';
      case Accidental.DBLFLAT:
        return 'bb';
      default:
        return '';
    }
  }

  export function toStringPretty(a: Accidental): string {
    switch (a) {
      case Accidental.NAT:
        return '';
      case Accidental.SHARP:
        return '#';
      case Accidental.DBLSHARP:
        return '𝄪';
      case Accidental.FLAT:
        return '♭';
      case Accidental.DBLFLAT:
        return '♭♭';
      default:
        return '';
    }
  }

  export function toABC(a: Accidental): string {
    switch (a) {
      case Accidental.NAT:
        return '=';
      case Accidental.SHARP:
        return '^';
      case Accidental.DBLSHARP:
        return '^^';
      case Accidental.FLAT:
        return '_';
      case Accidental.DBLFLAT:
        return '__';
      default:
        return '';
    };
  }

  export function fromString(s: string): Accidental {
    switch (s) {
      case '':
        return Accidental.NAT;
      case '#':
        return Accidental.SHARP;
      case '##':
        return Accidental.DBLSHARP;
      case 'b':
        return Accidental.FLAT;
      case 'bb':
        return Accidental.DBLFLAT;
      default:
        return Accidental.NAT;
    }
  }
}
export default Accidental;

File: server/src/main/java/edu/brown/cs/counterpointer/notenotation/pitch/Accidental.java
package edu.brown.cs.counterpointer.notenotation.pitch;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;
import edu.brown.cs.counterpointer.notenotation.exceptions.TextNotFoundException;

/**
 * Represents an accidental, which changes the note by 0, 1, or 2 half steps, up or down.
 */
public enum Accidental {
  DOUBLE_FLAT(-2, "bb"),
  FLAT(-1, "b"),
  NATURAL(0, ""),
  SHARP(1, "#"),
  DOUBLE_SHARP(2, "##");

  private final int valueChange;
  private final String abcjsText;

  /**
   * Enum constructor.
   * @param valueChange the number of half steps by which this accidental changes a note
   * @param abcjsText Text representation
   */
  Accidental(int valueChange, String abcjsText) {
    this.valueChange = valueChange;
    this.abcjsText = abcjsText;
  }

  /**
   * Getter for valueChange.
   * @return the number of half steps by which this accidental changes a note
   */
  @JsonValue
  public int getValueChange() {
    return this.valueChange;
  }

  @JsonCreator
  public static Accidental fromValueChange(int valueChange) {
    for (Accidental a : values()) {
      if (a.valueChange == valueChange) {
        return a;
      }
    }
    throw new IllegalArgumentException("Invalid valueChange " + valueChange);
  }

  /**
   * Getter for text.
   * @return the text for this accidental
   */
  public String getAbcjsText() {
    return abcjsText;
  }

  /**
   * Gets the accidental represented by the string.
   * @param s the string representation of a Accidental
   * @return the Accidental represented by s
   */
  public static Accidental byString(String s) {
    for (Accidental a : Accidental.values()) {
      if (a.getAbcjsText().equals(s)) {
        return a;
      }
    }
    throw new TextNotFoundException();
  }

  @Override
  public String toString() {
    return abcjsText;
  }
}
